src/lib/marshal/email-layout/writer.ts

```diff
@@ -0,0 +1,193 @@
+        : fs.outputFile(filePath, fileContent);
+    });
+
+    await Promise.all(promises);
+  } catch (error) {
+    // In case of any error, wipe the target directory that is likely in a bad
+    // state then restore the backup if one existed before.
+    if (emailLayoutDirCtx.exists) {
+      await fs.emptyDir(emailLayoutDirCtx.abspath);
+      await fs.copy(backupDirPath, emailLayoutDirCtx.abspath);
+    } else {
+      await fs.remove(emailLayoutDirCtx.abspath);
+    }
+
+    throw error;
+  } finally {
+    // Always clean up the backup directory in the temp sandbox.
+    await fs.remove(backupDirPath);
+  }
+};
+
+/* For a given email layout payload, this function builds a "email layout directoy bundle".
+   This is an object which contains all the relative paths and its file content.
+   It includes the extractable fields, which are extracted out and added to the bundle as separate files.
+*/
+const buildEmailLayoutDirBundle = (
+  emailLayout: EmailLayoutData<WithAnnotation>,
+): EmailLayoutDirBundle => {
+  const bundle: EmailLayoutDirBundle = {};
+
+  // A map of extraction settings of every field in the email layout
+  const compiledExtractionSettings = compileExtractionSettings(emailLayout);
+
+  // Iterate through each extractable field, determine whether we need to
+  // extract the field content, and if so, perform the
+  // extraction.
+  for (const [objPath, extractionSettings] of compiledExtractionSettings) {
+    const objPathStr = ObjPath.stringify(objPath);
+    const { default: extractByDefault, file_ext: fileExt } = extractionSettings;
+
+    if (!extractByDefault) continue;
+    // By this point, we have a field where we need to extract its content.
+    const data = get(emailLayout, objPath);
+    const fileName = objPath.pop();
+    const relpath = `${fileName}.${fileExt}`;
+
+    // Perform the extraction by adding the content and its file path to the
+    // bundle for writing to the file system later. Then replace the field
+    // content with the extracted file path and mark the field as extracted
+    // with @ suffix.
+    set(bundle, [relpath], data);
+    set(emailLayout, `${objPathStr}${FILEPATH_MARKER}`, relpath);
+    unset(emailLayout, objPathStr);
+  }
+
+  // At this point the bundle contains all extractable files, so we finally add the layout
+  // JSON realtive path + the file content.
+
+  return set(bundle, [LAYOUT_JSON], toEmailLayoutJson(emailLayout));
+};
+
+// This bulk write function takes the fetched email layouts data KNOCK API and writes
+// them into a directory.
+export const writeEmailLayoutIndexDir = async (
+  indexDirCtx: DirContext,
+  emailLayouts: EmailLayoutData<WithAnnotation>[],
+): Promise<void> => {
+  const backupDirPath = path.resolve(sandboxDir, uniqueId("backup"));
+
+  try {
+    if (indexDirCtx.exists) {
+      await fs.copy(indexDirCtx.abspath, backupDirPath);
+      await fs.emptyDir(indexDirCtx.abspath);
```

Comment:
If we always blow out everything in the index directory, then we won't be able to preserve the extracted file paths used in the existing email layout dirs. Could we take the approach similar to workflows where we selectively remove any email layouts that aren't found in `emailLayouts`?